### PR TITLE
HWKAPM-590 Move the Property set from Trace to Node, to enable better…

### PR DESCRIPTION
**Pull request by objectiser**
… correlation between the binary annotations/tags reported on Spans in zipkin/opentracing

### CODE REVIEW

**Comment by objectiser** [comment]
@jpkrohling Needs further work and testing, but wanted to get your thoughts on first draft.

Reviewer notes:
* The current model associates 'searchable' properties (generally intended for business relevant context information) with the Trace class - which means collected for each fragment but then aggregated into the single overall end to end Trace
* These properties also get associated with most/all of the derived information created from the collected fragments aswell
* With the recent work on aligning with zipkin and opentracing style trace models, their models have 'tags' or 'binary annotations' associated with spans, which are equivalent to the APM nodes
* So to ensure we capture information at the same level of granularity, it seems more appropriate to store the APM properties on the nodes, rather than the Trace fragments
* Issue is that to make the properties on the fragments searchable in Elasticsearch, they need to be aggregated and made available at the trace level. This has been accomplished using custom serialiser/deserialiser.

**Comment by objectiser** [comment]
Tested against swarm booker example and ticketmonster app. Both extract business properties, which are still handled in the same way by the server, even though now stored on the nodes rather than trace fragments.

File: api/src/main/java/org/hawkular/apm/api/model/trace/Node.java [hunk 1]
```diff
@@ -213,6 +218,20 @@ public Node setFaultDescription(String faultDescription) {
     }
 
     /**
+     * @return the properties
+     */
+    public Set<Property> getProperties() {
+        return properties;
```
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/model/trace/Node.java]: Do you intend to allow external objects to manipulate the contents of this set directly? If not, then it might be better to wrap the Set into a `Collections.unmodifiableSet()`

File: api/src/main/java/org/hawkular/apm/api/model/trace/Node.java [hunk 2]
```diff
@@ -213,6 +218,20 @@ public Node setFaultDescription(String faultDescription) {
...
+    }
+
+    /**
+     * @param properties the properties to set
+     */
+    public void setProperties(Set<Property> properties) {
```
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/model/trace/Node.java]: Similarly, you might want to create a new Set based on the set you received, to ensure the contents aren't changed by some external object.

File: api/src/main/java/org/hawkular/apm/api/model/trace/Node.java [hunk 3]
```diff
@@ -231,6 +250,49 @@ public Node setDetails(Map<String, String> details) {
...
 
     /**
+     * This method determines whether there is atleast one
+     * property with the supplied name.
+     *
+     * @param name The property name
+     * @return Whether a property of the supplied name is defined
+     */
+    public boolean hasProperty(String name) {
+        for (Property property : this.properties) {
```
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/model/trace/Node.java]: If Java 8 can be used, this method can be reduced to:
`return this.properties.stream().filter(p -> p.getName().equals(name)).findFirst().isPresent();`

File: api/src/main/java/org/hawkular/apm/api/model/trace/ContainerNode.java [hunk 4]
```diff
@@ -74,6 +76,20 @@ public void setNodes(List<Node> nodes) {
...
     /**
+     * This method adds the properties for this node to the
+     * supplied set.
+     *
+     * @param allProperties The aggregated set of properties
+     */
+    @Override
+    protected void includeProperties(Set<Property> allProperties) {
+        super.includeProperties(allProperties);
+        for (Node n : nodes) {
```
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/model/trace/ContainerNode.java]: If Java 8 is allowed: `nodes.forEach(n -> n.includeProperties(allProperties));`

**Comment by jpkrohling** [comment]
Before I continue with more Java 8 related comments, let me say that those can be safely ignored in most of the cases :) My personal taste is to use them, but I understand that Java 8 might not be desirable, or that your taste is different.

File: api/src/main/java/org/hawkular/apm/api/model/trace/Node.java [hunk 5]
```diff
@@ -231,6 +250,49 @@ public Node setDetails(Map<String, String> details) {
...
+
+    /**
+     * This method returns the set of properties having the
+     * supplied property name.
+     *
+     * @param name The property name
+     * @return The set of properties with the supplied name
+     */
+    public Set<Property> getProperties(String name) {
+        Set<Property> ret = new HashSet<Property>();
```
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/model/trace/Node.java]: java 8: return `this.properties.stream().filter(property -> property.getName().equals(name)).collect(Collectors.toSet());`

File: api/src/main/java/org/hawkular/apm/api/model/trace/Node.java (hunk 2 above)
Comment (jpkrohling) [inline api/src/main/java/org/hawkular/apm/api/model/trace/Node.java]: Also, the caller is able to set the properties to `null`. If that's acceptable, can you add a simple Javadoc stating this fact?

File: api/src/test/java/org/hawkular/apm/api/model/trace/TraceTest.java [hunk 6]
```diff
@@ -260,4 +259,26 @@ public void testCalculateDuration() {
...
+
+    @Test
+    public void testAllProperties() {
+        Trace trace1 = new Trace();
+        trace1.setId("1");
+        trace1.setBusinessTransaction("testapp");
+
+        Consumer c1 = new Consumer();
+        c1.getProperties().add(new Property("prop1", "value1"));
+        trace1.getNodes().add(c1);
```
Comment (jpkrohling) [inline api/src/test/java/org/hawkular/apm/api/model/trace/TraceTest.java]: I think this answers a question I had above, about the external object having the ability to manipulate the list directly

File: client/collector/src/main/java/org/hawkular/apm/client/collector/DefaultTraceCollector.java [hunk 7]
```diff
@@ -856,19 +856,16 @@ public void setFault(String location, String value, String description) {
...
         try {
             if (fragmentManager.hasFragmentBuilder()) {
                 FragmentBuilder builder = fragmentManager.getFragmentBuilder();
 
-                if (value == null) {
-                    builder.getTrace().getProperties().remove(name);
-                } else {
-                    builder.getTrace().getProperties().add(new Property(name, value));
-                }
+                builder.getCurrentNode().getProperties().add(new Property(name, value));
```
Comment (jpkrohling) [inline client/collector/src/main/java/org/hawkular/apm/client/collector/DefaultTraceCollector.java]: If the `value` is null, the property will be updated to have its value as `null`, while previously the semantics were `remove the entry if the value is null`. Is this what you had in mind on this change?

File: api/src/main/java/org/hawkular/apm/api/model/trace/Node.java (hunk 1 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/model/trace/Node.java]: Yes, as seen further down, properties can be added over time.

File: api/src/main/java/org/hawkular/apm/api/model/trace/Node.java (hunk 2 above)
Comment (objectiser) [inline api/src/main/java/org/hawkular/apm/api/model/trace/Node.java]: Null shouldn't be allowed, so will update.

File: client/collector/src/main/java/org/hawkular/apm/client/collector/DefaultTraceCollector.java (hunk 7 above)
Comment (objectiser) [inline client/collector/src/main/java/org/hawkular/apm/client/collector/DefaultTraceCollector.java]: The previous code would not work (a bug) as it is trying to remove a string from a set of Property objects. This was a hang over from when getProperties() was a map.

However properties are only added by the collector, and it is possible a property may want to represent a null value, so felt it was better to simplify the code.

**Comment by objectiser** [comment]
@jpkrohling Thanks for the review. In terms of the java8 changes - I'll have a look at them and decide which ones to possibly adopt.

**Comment by objectiser** [comment]
@jpkrohling Changes applied - let me know if there is anything else. Thanks.

File: processors/communicationdetails-deriver/src/test/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriverTest.java [hunk 8]
```diff
@@ -473,11 +473,11 @@ public void testProcessSingleWithClient() {
         trace2.setHostName("host2");
         trace2.setHostAddress("addr2");
         trace2.setPrincipal("p1");
-        trace2.getProperties().add(new Property("prop1", "value1"));
 
         Consumer c2 = new Consumer();
         c2.setUri("TheURI");
         c2.setDuration(1200000000);
+        c2.getProperties().add(new Property("prop1", "value1"));
```
Comment (jpkrohling) [inline processors/communicationdetails-deriver/src/test/java/org/hawkular/apm/processor/communicationdetails/CommunicationDetailsDeriverTest.java]: Can you also add a test with a Property with a null value, to test the condition mentioned on a previous comment?

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java [hunk 9]
```diff
@@ -314,4 +347,84 @@ public void clear(String tenantId) {
...
+                throws IOException, JsonProcessingException {
+            jgen.writeStartObject();
+            jgen.writeStringField(BUSINESS_TRANSACTION_FIELD, trace.getBusinessTransaction());
+            jgen.writeStringField(HOST_ADDRESS_FIELD, trace.getHostAddress());
+            jgen.writeStringField(HOST_NAME_FIELD, trace.getHostName());
+            jgen.writeStringField(ID_FIELD, trace.getId());
+            jgen.writeStringField(PRINCIPAL_FIELD, trace.getPrincipal());
+            jgen.writeNumberField(START_TIME_FIELD, trace.getStartTime());
+            jgen.writeArrayFieldStart(NODES_FIELD);
+            for (Node n : trace.getNodes()) {
```
Comment (jpkrohling) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: You could use a `forEach` here as well: `trace.getNodes().forEach(jgen::writeObject)`

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java [hunk 10]
```diff
@@ -314,4 +347,84 @@ public void clear(String tenantId) {
...
+                jgen.writeObject(n);
+            }
+            jgen.writeEndArray();
+            Set<Property> properties = trace.allProperties();
+            jgen.writeArrayFieldStart(PROPERTIES_FIELD);
+            for (Property p : properties) {
```
Comment (jpkrohling) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: Ditto

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java [hunk 11]
```diff
@@ -314,4 +347,84 @@ public void clear(String tenantId) {
...
+        }
+    }
+
+    public static class TraceDeserializer extends JsonDeserializer<Trace> {
+
+        /* (non-Javadoc)
+         * @see com.fasterxml.jackson.databind.JsonDeserializer#deserialize(com.fasterxml.jackson.core.JsonParser, com.fasterxml.jackson.databind.DeserializationContext)
+         */
+        @Override
+        public Trace deserialize(JsonParser parser, DeserializationContext context)
```
Comment (jpkrohling) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: Does Jackson have problems in automatically serializing/deserializing Traces? If so, would be worth adding the reason as a javadoc

File: tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java [hunk 12]
```diff
@@ -328,15 +329,12 @@ public void testGetCompletionCountWithPropertyFilter() {
...
         trace1.getNodes().add(c1);
-        trace1.getProperties().add(new Property("prop1", "2.5", PropertyType.Number));
-        trace1.getProperties().add(new Property("prop2", "hello"));
-
-        List<Trace> traces = new ArrayList<Trace>();
-        traces.add(trace1);
 
         try {
-            publisher.publish(null, traces);
+            publisher.publish(null, Arrays.asList(trace1));
```
Comment (jpkrohling) [inline tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/AnalyticsServiceRESTTest.java]: Perhaps it has the same effect, but I have the feeling that a `Collections.singletonList()` would be more appropriate. In any case, feel free to ignore this comment, as this is test code :)

File: tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/TraceServiceRESTTest.java [hunk 13]
```diff
@@ -164,12 +165,8 @@ public void testStoreAndRetrieveComplexTraceById() {
         p2_2.addInteractionId("id2_2");
         c2.getNodes().add(p2_2);
 
-        List<Trace> traces = new ArrayList<Trace>();
-        traces.add(trace1);
-        traces.add(trace2);
-
         try {
-            publisher.publish(null, traces);
+            publisher.publish(null, Arrays.asList(trace1, trace2));
```
Comment (jpkrohling) [inline tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/TraceServiceRESTTest.java]: Does it make sense to add an alternate `publish` for single traces? It seems it's far more common now than before?

**Comment by jpkrohling** [comment]
Review finished. Nothing serious was found though, I'd be OK if it were merged as it is.

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java (hunk 11 above)
Comment (objectiser) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: As mentioned in the reviewers notes, the reason for explicitly using custom serializer/deserializer is to modify the json stored in Elasticsearch, to aggregate the properties associated with the node hierarchy and associate them with the top level Trace. This is to enable them to be queried, as Elasticsearch doesn't support querying fields on a recursive structure.

File: server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java (hunk 9 above)
Comment (objectiser) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: Tried it, and also trace.getNodes().forEach(n -> jgen.writeObject(n));, but geting "Unhandled exception type: IOException" in Eclipse (and also maven from command line) - even when a try/catch(Throwable) block surrounds.

Any ideas? Using jackson 2.6.0.

Comment (objectiser) [inline server/elasticsearch/src/main/java/org/hawkular/apm/server/elasticsearch/TraceServiceElasticsearch.java]: Tried with latest 2.8.1 and got the same problem.

File: tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/TraceServiceRESTTest.java (hunk 13 above)
Comment (objectiser) [inline tests/server/wildfly/src/test/java/org/hawkular/apm/tests/wildfly/TraceServiceRESTTest.java]: Only common in tests :)

**Comment by objectiser** [comment]
@jpkrohling Thanks Juca, I've updated the tests, but will leave the serializer as having strange compilation issues when changed to java8.

**Comment by jpkrohling** [comment]
Sure, looks good to me!
